fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/PersistentDynamicRegistryHandler.java

```diff
@@ -0,0 +1,200 @@
+		return registries;
+	}
+
+	/**
+	 * Remaps a registry if existing data is passed in.
+	 * Then writes out the ids in the registry (remapped or a new world).
+	 * Keeps hold of the orphaned registry entries as to not overwrite them.
+	 */
+	private static <T> CompoundTag remapRegistry(Identifier registryId, MutableRegistry<T> registry, @Nullable CompoundTag existingTag) throws RemapException {
+		if (!(registry instanceof RemappableRegistry)) {
+			throw new UnsupportedOperationException("Cannot remap un re-mappable registry: " + registryId.toString());
+		}
+
+		boolean isModded = registry.getIds().stream().anyMatch(id -> !id.getNamespace().equals("minecraft"));
+
+		// The current registry might not be modded, but we might have previous changed vanilla ids that we should try and remap
+		if (existingTag != null && !isModded) {
+			isModded = existingTag.getKeys().stream()
+					.map(existingTag::getString)
+					.map(Identifier::new)
+					.anyMatch(id -> !id.getNamespace().equals("minecraft"));
+		}
+
+		if (LOGGER.isDebugEnabled()) {
+			if (existingTag == null) {
+				LOGGER.debug("No existing data found, assuming new registry with {} entries. modded = {}", registry.getIds().size(), isModded);
+			} else {
+				LOGGER.debug("Existing registry data found. modded = {}", isModded);
+
+				for (T entry : registry) {
+					//noinspection unchecked
+					Identifier id = registry.getId(entry);
+					int rawId = registry.getRawId(entry);
+
+					if (id == null) continue;
+
+					if (existingTag.getKeys().contains(id.toString())) {
+						int existingRawId = existingTag.getInt(id.toString());
+
+						if (rawId != existingRawId) {
+							LOGGER.debug("Remapping {} {} -> {}", id.toString(), rawId, existingRawId);
+						} else {
+							LOGGER.debug("Using existing id for {} {}", id.toString(), rawId);
+						}
+					} else {
+						LOGGER.debug("Found new registry entry {}", id.toString());
+					}
+				}
+			}
+		}
+
+		// If we have some existing ids and the registry contains modded/datapack entries we remap the registry with those
+		if (existingTag != null && isModded) {
+			LOGGER.debug("Remapping {} with {} entries", registryId, registry.getIds().size());
+			Object2IntMap<Identifier> idMap = new Object2IntOpenHashMap<>();
+
+			for (String key : existingTag.getKeys()) {
+				idMap.put(new Identifier(key), existingTag.getInt(key));
+			}
+
+			((RemappableRegistry) registry).remap(registryId.toString(), idMap, RemappableRegistry.RemapMode.AUTHORITATIVE);
+		} else {
+			LOGGER.debug("Skipping remap of {}", registryId);
+		}
+
+		// Now start to build up what we are going to save out
+		CompoundTag registryTag = new CompoundTag();
+
+		// Save all ids as they appear in the remapped, or new registry to disk even if not modded.
+		for (T entry : registry) {
+			//noinspection unchecked
+			Identifier id = registry.getId(entry);
+
+			if (id == null) {
+				continue;
+			}
+
+			int rawId = registry.getRawId(entry);
```

Comment:
Do we want to validate rawId too? (skip < 0)

Reply:
I think the registry is gonna give you -1 if the object is not keyed in registry

Reply:
Good idea, might as well.